### PR TITLE
refactor addon controller to support disable old and enable new installation

### PR DESCRIPTION
**Pull request by zhiweiyin318**
story: https://github.com/stolostron/backlog/issues/19282

Signed-off-by: Zhiwei Yin <zyin@redhat.com>

### CODE REVIEW

File: pkg/controller/addon/klusterlet_addon_controller.go [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+
+func (r *ReconcileKlusterletAddOn) deleteManagedClusterAddon(addonName, clusterName string) error {
+	if !agentv1.KlusterletAddons[addonName] {
+		return nil
+	}
+	if addonName == agentv1.WorkManagerAddonName {
+		return nil
+	}
+	addon := &addonv1alpha1.ManagedClusterAddOn{}
+	err := r.client.Get(context.TODO(), types.NamespacedName{Name: addonName, Namespace: clusterName}, addon)
```
Comment (elgnay) [inline pkg/controller/addon/klusterlet_addon_controller.go:224]: I think you don't have to `get` the addon before deleting it.
https://github.com/kubernetes-sigs/controller-runtime/blob/273e608d8a6cd7f34b6b56a6bd4626878319aed7/pkg/client/example_test.go#L219

File: pkg/controller/addon/klusterlet_addon_controller.go [hunk 2]
```diff
@@ -0,0 +1,543 @@
...
+		if err != nil {
+			return err
+		}
+		if len(valuesString) != 0 {
+			newAddon.SetAnnotations(map[string]string{annotationValues: valuesString})
+		}
+
+		if err := r.client.Create(context.TODO(), newAddon); err != nil {
+			return err
+		}
```
Comment (elgnay) [inline pkg/controller/addon/klusterlet_addon_controller.go:259]: Set `addon = newAddon` once addon is created?

File: pkg/utils/utils.go [hunk 3]
```diff
@@ -199,7 +204,12 @@ func CreateOrUpdateManifestWork(
 	if err == nil {
 		// Check if update is require
 		if !compareManifestWorks(&oldManifestwork, manifestwork) {
-			oldManifestwork.Spec.Workload.Manifests = manifestwork.Spec.Workload.Manifests
+			oldManifestwork.Spec = manifestwork.Spec
+			if oldManifestwork.Labels == nil {
+				oldManifestwork.Labels = map[string]string{agentv1.UpgradeLabel: ""}
```
[the remark this answers is not in the document]
Comment (elgnay) [reply pkg/utils/utils.go:209]: How about user changes the manifests included in the spec? Will the update label been added in this case?

File: pkg/controller/addon/klusterlet_addon_controller.go (hunk 1 above)
Comment (zhiweiyin318) [reply pkg/controller/addon/klusterlet_addon_controller.go]: done

File: pkg/controller/addon/klusterlet_addon_controller.go (hunk 2 above)
Comment (zhiweiyin318) [reply pkg/controller/addon/klusterlet_addon_controller.go]: changed to` return r.client.Create(context.TODO(), newAddon). `

**Pull request merged**